File: VSCode/Python/Algorithm/group_anagrams.py
# Anagram: 문자를 재배열하여 다른 뜻을 가진 단어로 바꾸는 것
import collections

def group_anagrams(words):
    anagrams = collections.defaultdict(list)
    
    for word in words:
        # 정렬하여 딕셔너리에 추가
        # sorted -> Timsort 기반 리스트를 반환
        # sort -> In-place, 리스트를 반환하지 않고 리스트 자체를 바꿈
        anagrams[''.join(sorted(word))].append(word)
    return list(anagrams.values())


if __name__ == "__main__":
    words = ["eat", "tea", "tan", "ate", "nat", "bat" ]
    print(group_anagrams(words))
    

File: Java/src/LeetCode/GroupAnagrams.java
package LeetCode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GroupAnagrams {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> anaMap = new HashMap<String, List<String>>();

        for(String str : strs){
            char[] toChars = str.toCharArray();
            Arrays.sort(toChars);
            String sortedStr = new String(toChars);
            if(anaMap.containsKey(sortedStr)){
                anaMap.get(sortedStr).add(str);
            } else{
                anaMap.put(sortedStr, new ArrayList<>());
                anaMap.get(sortedStr).add(str);
            }
            //System.out.println("Key : " + sortedStr + ", Values : " + anaMap.get(sortedStr).toString());
        }

        List<List<String>> answer = new ArrayList<List<String>>();
        for(var iter : anaMap.entrySet()){
            answer.add(iter.getValue());
        }
        return answer;
    }

    public static void main(String[] args){
        new GroupAnagrams().groupAnagrams(new String[]{"eat","tea","tan","ate","nat","bat"});
    }
}
